### PR TITLE
add policy enforcement to volume operations

### PR DESCRIPTION
**Pull request by AmitKumarDas**
1. Why is this change necessary ?

It has been seen that policy enforcement is required
for OpenEBS volumes during its provisioning operations.
A volume policy can be related to placements, security/backup,
workload characteristics, QOS, etc.

Policy Enforcement Workflow: A volume provisioning request
to maya is intercepted by volume_endpoint.go file. This is
made to pass through various policy enforcement logic. The
end point decides on the volume provisioner based on VolumeType.
It is then directed to the selected provisioner. The provisioner
then decides on the volume orchestrator provider & directs the
request to the selected orchestrator. The orchestrator in turn
is responsible to convert the volume provisioning request to its
own structure and pass it to its APIs.

This particular commit does a partial
fix of openebs/openebs#409, openebs/openebs#721,
openebs/openebs#749, openebs/openebs#751

2. How does this change address the issue ?

- Add & Delete volume goes through policy enforcement logic

- More properties & setters are introduced in K8s structures to
  handle state & in turn make them unit-testable

- Functions have been introduced to create new instances of
  certain structures. This will remove the need for fetching
  instances via registration

- Policy related operations contract has been introduced in
  orchestrator interface

- A new, simple but redundant wrapper over K8s client-go
  operations has been introduced. This avoids modification of
  existing code while caters to new requirements. This will
  be used instead of older wrapper in near future.

- Typed structures have been introduced instead of
  map[string]string

3. How to verify this change ?

- make should work fine

- a set of issues will be created that verifies
  policy enforcement via unit tests, e2e tests

4. What side effects does this change have ?

- Testing is required to un-earth side effects

- Lack of unit tests is a concern

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: orchprovider/k8s/v1/k8s.go [hunk 1]
```diff
@@ -117,23 +151,54 @@ func k8sOrchUtil(k *k8sOrchestrator, volProfile volProfile.VolumeProvisionerProf
...
+	sc, err := scOps.Get(k.volume.Labels.K8sStorageClass, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return sc.Parameters, nil
+}
+
 // AddStorage will add persistent volume running as containers. In OpenEBS
 // terms AddStorage will add a VSM.
```
Comment (prateekpandey14) [inline orchprovider/k8s/v1/k8s.go]: will as a Volume ?

File: orchprovider/k8s/v1/k8s.go [hunk 2]
```diff
@@ -1100,7 +1165,7 @@ func (k *k8sOrchestrator) getReplicaPods(vsm string, podOps k8sCoreV1.PodInterfa
 	return rp, nil
 }
 
-// getPods deletes the Pods w.r.t the VSM
+// getPods gets the Pods w.r.t the VSM
```
Comment (prateekpandey14) [inline orchprovider/k8s/v1/k8s.go]: same as above (Volume)

File: orchprovider/k8s/v1/k8s_test.go [hunk 3]
```diff
@@ -176,17 +178,17 @@ func (m *mockK8sOrch) GetK8sUtil(volProfile volProfile.VolumeProvisionerProfile)
...
-		ns:                 vol.Labels[string(v1.OrchNSLbl)],
-		injectNSErr:        vol.Labels[string(testK8sInjectNSErrLbl)],
-		inCluster:          vol.Labels[string(testK8sInClusterLbl)],
-		injectInClusterErr: vol.Labels[string(testK8sInjectInClusterErrLbl)],
-		injectPodErr:       vol.Labels[string(testK8sInjectPodErrLbl)],
-		injectSvcErr:       vol.Labels[string(testK8sInjectSvcErrLbl)],
-		injectVsm:          vol.Labels[string(testK8sInjectVSMLbl)],
-		resultingErr:       vol.Labels[string(testK8sErrorLbl)],
+		//name:               vol.Labels[string(testK8sUtlNameLbl)],
+		vsmName: vol.Name,
```
Comment (prateekpandey14) [inline orchprovider/k8s/v1/k8s_test.go]: can we change this struct field vsmName ---> volName

File: orchprovider/k8s/v1/k8s.go (hunk 1 above)
Comment (AmitKumarDas) [reply orchprovider/k8s/v1/k8s.go]: @prateekpandey14 Did not understand the query

File: orchprovider/k8s/v1/k8s.go (hunk 2 above)
Comment (AmitKumarDas) [reply orchprovider/k8s/v1/k8s.go]: Ok. These will be refactored to volume_add.go & so on files. Will be doing these renaming stuff if it exists after all these refactorings.

File: orchprovider/k8s/v1/k8s_test.go (hunk 3 above)
Comment (AmitKumarDas) [reply orchprovider/k8s/v1/k8s_test.go]: YUP. We will change. Not now. There are a lot of refactoring planned with unit tests.

File: orchprovider/k8s/v1/k8s.go (hunk 1 above)
Comment (prateekpandey14) [reply orchprovider/k8s/v1/k8s.go]: i.e. VSM --> Volume

**Pull request closed without merging**